### PR TITLE
Separate SSLSocket from Socket

### LINKED ISSUE #4057: ESP32-S2 "Out of sockets" in 6.2.0-beta.0
Context manager __exit__ seems to be in place for socket, but:
```
import time
import wifi
import socketpool

from secrets import secrets

pool = socketpool.SocketPool(wifi.radio)

print("Connect...")
wifi.radio.connect(secrets["ssid"], secrets["password"])
print("RSSI", wifi.radio.ap_info.rssi)

count = 0
packet = bytearray(1)
while True:
    count += 1
    print(count, end=" ")
    packet[0] = 0xFF
    with pool.socket(pool.AF_INET, pool.SOCK_DGRAM) as sock:
        print("Send...")
        sock.sendto(packet, ("192.168.4.32", 16384))
    time.sleep(1)
```
results in:
```
Adafruit CircuitPython 6.2.0-beta.0 on 2021-01-22; FeatherS2 with ESP32S2
>>> import ntp
Connect...
RSSI -43
1 Send...
2 Send...
3 Send...
4 Send...
5 Send...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "ntp.py", line 21, in <module>
  File "ntp.py", line 21, in <module>
RuntimeError: Out of sockets
```
CPython version continues past where my patience wears out:
```
import time
import socket

pool = socket

count = 0
packet = bytearray(1)
while True:
    count += 1
    print(count, end=" ")
    packet[0] = 0xFF
    with pool.socket(pool.AF_INET, pool.SOCK_DGRAM) as sock:
        print("Send...")
        sock.sendto(packet, ("192.168.4.32", 16384))
    time.sleep(1)
```
6.1.0-rc.1 seems to behave fine.

May be related to [https://github.com/adafruit/Adafruit_CircuitPython_Requests/issues/63](https://github.com/adafruit/Adafruit_CircuitPython_Requests/issues/63)

### LINKED ISSUE #4101: SocketPool.Socket.__hash__() is not immutable
`__hash__()` was added to `Socket` so a `Socket` could be put in a dict, for use in the `adafruit_requests` library. However, the hash used is not immutable, and changes when the socket is closed.

The hash returned could be `id()`, to make it immutable.. However, it might be worth looking at whether the use of `Socket` in `adafruit_requests` really needs a dictionary. I see two uses: `._open_sockets` and `._socket_free`. I'll bring this up in an issue in that repo.

Tagging @hierophect on this, who was trying to debug #4061, which may be caused by this. (`SSLSocket` has the same hash.)

### PR DESCRIPTION
**Pull request by hierophect**
This PR removes the TLS implementations from Socket and moves them to a new object class, SSLSocket, which is created from a Socket object using SSL's `contex.wrap_socket()` function. This behavior more closely matches that of [Cpython](https://docs.python.org/3/library/ssl.html#ssl-sockets).

Opening as a draft as I've had a couple difficulties with this change and I'm still testing:

- The regular Socket class has a new LWIP-only implementation for `connect`, but I've had issues making this non-blocking. A non-blocking call to `connect` should always return a -1 "EINPROGRESS" error for the first call, but I don't find that they *ever* connect, no matter how long the timeout is. The current implementation simply switches the entire socket to blocking mode briefly, but other failed implementations are shown as comments. 
- Still trying to figure out exactly how SSLSocket implements `bind` and `listen` and `accept` - as best I can tell, they should call the original functions (ie LWIP ones) and then just wrap whatever socket gets accepted as a SSLSocket. But I'm still not sure about this.
- I don't have a great SSL python example so if anyone has a demo they like that'd be good.

Also includes bugfixes for a variety of user-reported issues with Socket.
Resolves #4057
Resolves #4101

### CODE REVIEW

File: ports/esp32s2/common-hal/ssl/SSLContext.c [hunk 1]
```diff
@@ -25,17 +25,35 @@
...
 
-    socket->ssl_context = self;
+    ssl_sslsocket_obj_t *sock = m_new_obj_with_finaliser(ssl_sslsocket_obj_t);
+    sock->base.type = &ssl_sslsocket_type;
+    sock->ssl_context = self;
+    sock->sock = socket;
+
+    if (socket->type != SOCK_STREAM || socket->num != -1) {
+        mp_raise_RuntimeError(translate("Invalid socket for TLS"));
+    }
```
Comment (tannewt) [inline ports/esp32s2/common-hal/ssl/SSLContext.c:48]: Do this before the allocation above.

**Review by tannewt — COMMENTED** [review]
Skimmed it and only see one thing to change. Thanks for following up with this!

**Comment by hierophect** [comment]
It should also go without saying that this PR breaks the heck out of [requests](https://github.com/adafruit/Adafruit_CircuitPython_Requests)

**Comment by tannewt** [comment]
> It should also go without saying that this PR breaks the heck out of [requests](https://github.com/adafruit/Adafruit_CircuitPython_Requests)

Why? It shouldn't because this is the same API as CPython which requests works with.

**Comment by hierophect** [comment]
> Why? It shouldn't because this is the same API as CPython which requests works with.
Yes, sorry, you're right. There was an error in `wrap_socket` that I confused for being an incompatibility. Wrap_socket now accepts sockets with a valid LWIP number since they're used for the bind-listen-accept methods.

**Review by tannewt — APPROVED** [review]
This looks good to me! Thank you for separating these out!

**Pull request merged**